clr/commands.py

```diff
@@ -494,7 +499,96 @@ def cmd_completion_arg(self, command_name, partial="", bools_only=False):
                 options.append(f"--no{param.name}")
         # partial is prepended with a space to stop argparse from parsing it
         partial = partial.strip()
-        print("\n".join(f"{o} " for o in options if o.startswith(partial)), end="")
+        print_for_complete(partial, results)
+
+    def cmd_smart_complete(self, *existing_args):
+        """Smart/opinionated completion.
+
+        Only completes the next missing required arg if there are any. Afterward
+        suggests optional args."""
+
+        if len(existing_args) < 2:
+            # Invalid call.
+            return
+
+        command_name = existing_args[1]
+        existing_args = existing_args[2:]
+
+        # existing_args will contain at least an empty str element if we're past the command name.
+        if not existing_args:
+            self.cmd_complete_command(query=command_name)
+            return
+
+        # Special case for the first/only arg to help.
+        if command_name in ("system:help", "help") and len(existing_args) < 2:
+            self.cmd_complete_command(query=existing_args[0])
+            return
+
+        current_arg = existing_args[-1]
+        previous_args = existing_args[:-1]
+
+        def is_positional(arg):
+            return not arg.startswith("--")
+
+        existing_positional_args = len(list(takewhile(is_positional, previous_args)))
+
+        namespace_key, command_name = resolve_command(command_name, cache=self.cache)
+        parameters = (
+            self.cache.get(namespace_key)
+            .command_specs[command_name]
+            .signature.parameters.values()
+        )
+
+        # Scan over all parameters for the command to build up data for the following purposes:
+        # Will suggest the first missing required arg if there is one.
+        missing_required_args = []
+        # Once all the required args are present, will suggest the optional ones.
+        missing_optional_args = []
+        # Boolean args don't need to be followed by an value.
+        boolean_options = set()
+        # No filename completion for numberical args.
+        numerical_options = set()
+        # Required args don't have named flags if there is a var positional.
+        has_var_positional = False
+        for param_index, param in enumerate(parameters):
+            if param.kind == param.VAR_POSITIONAL:
+                has_var_positional = True
+                continue
+
+            required = param.default == Signature.empty
+            missing_args = missing_required_args if required else missing_optional_args
+            arg_names = [f"--{param.name}"]
+
+            if type(param.default) == bool:
+                arg_names.append(f"--no{param.name}")
```

Comment:
For my clarity, is this going to make params `--noparam` or `--no-param`? Seems like it'll be the prior. Any objections to doing the `--no-` prefix?

Reply:
`--noparam`

Would also prefer `--no-param` if we were starting from scratch, but this is done to match the old parsing scheme.

Reply:
Also, this PR is just the shell completion and doesn't actually define the flags.